### DISCUSSION TITLE
Able to define boundary representation? or surface from polyline?

### DISCUSSION THREAD

**Discussion by nimafo** [pyvista/pyvista#3904] [Q&A]
Hi all, 
Given all the points defining this combination, I need to find a way to define a surface with a hole on it. 

Here is an example of the outline surface: 
```
a = np.array([[0,0,0],[2,0,0],[2,2,0],[1.9,2,0],[1.9,1.9,0],[1.9,1,0],[0.5,1,0],[0.5,1.9,0],[1.9,1.9,0],[1.9,2,0],[0,2,0],[0,0,0]])
l = pv.lines_from_points(a)
l.plot()
```
resulting in: 
![Screenshot 2023-01-30 at 19 15 50](https://user-images.githubusercontent.com/45439717/215561270-106f5b26-2b4e-46d6-899d-3990de8bfd38.png)
Ideally, I want the rectangle in the center to be the hole, surrounded by a planar surface, represented by the boundaries above. 

Not irrelevant to this, I would also like to know whether it is possible to model[ boundary representations(b-reps)](https://en.wikipedia.org/wiki/Boundary_representation) in PyVista, by having the polyline boundary around that. 

Best,
N.

**Comment by akaszynski** [comment]
Here's how you'd do it with your geometry using the approach from [delaunay_2d](https://docs.pyvista.org/api/core/_autosummary/pyvista.UnstructuredGridFilters.delaunay_2d.html).

```py
import pyvista as pv

outer_box = pv.PolyData(
    [
        [0, 2, 0],
        [2, 2, 0],
        [2, 0, 0],
        [0, 0, 0],
    ],
    lines=[5, 0, 1, 2, 3, 0],
)

inner_box = pv.PolyData(
    [[1.9, 1.9, 0], [1.9, 1, 0], [0.5, 1, 0], [0.5, 1.9, 0]],
    lines=[5, 0, 1, 2, 3, 0],
    faces=[5, 0, 1, 2, 3, 0],
)

comb = inner_box + outer_box
out = comb.delaunay_2d(edge_source=comb)
out.plot(cpos="xy", show_edges=True)
```


![tmp](https://user-images.githubusercontent.com/11981631/215616728-bd6af7b1-e52a-4838-9b39-8eb471e30204.png)


> Not irrelevant to this, I would also like to know whether it is possible to model[ boundary representations(b-reps)](https://en.wikipedia.org/wiki/Boundary_representation) in PyVista, by having the polyline boundary around that.Not irrelevant to this, I would also like to know whether it is possible to model[ boundary representations(b-reps)](https://en.wikipedia.org/wiki/Boundary_representation) in PyVista, by having the polyline boundary around that.

I did something like this with https://github.com/pyvista/pyiges. It's not great, but it works for basic geometry. I think your best bet is using https://github.com/CadQuery/cadquery and then outputting it to PyVista to view it if needed.

**Reply by nimafo** [reply]
Hi @akaszynski,
This is an interesting solution,
but my data only has the point coordinates in order as a list and your code requires an additional split step between the inner points and outer points. 
This is an example: 
![Screenshot-2023-02-10-at-14 05](https://user-images.githubusercontent.com/45439717/218099767-6bd7aef4-ac29-47f7-b7ff-043611705f88.jpg)
![Screenshot-2023-02-10-at-14 05_1](https://user-images.githubusercontent.com/45439717/218099797-a3665c98-d698-4226-b869-9136aaac5010.jpg)
